cfme/services/requests.py

```diff
@@ -1,288 +1,498 @@
 
-REQUEST_FINISHED_STATES = {'Migrated', 'Finished'}
-
-buttons = Region(
-    locators=dict(
-        approve="//*[@title='Approve this Request']",
-        deny="//*[@title='Deny this Request']",
-        copy="//*[@title='Copy original Request']",
-        edit="//*[@title='Edit the original Request']",
-        delete="//*[@title='Delete this Request']",
-        submit="//span[@id='buttons_on']/a[@title='Submit']",
-        cancel="//a[@title='Cancel']",
-    )
-)
-
-request_table = Table('//*[@id="list_grid"]/table')
-fields = Region(
-    locators=dict(
-        reason=Input("reason"),
-        request_list=request_table
-    )
-)
 
-match_page = partial(match_location, controller='miq_request', title='Requests')
+class Request(Navigatable):
+    """
+    Class describes request row from Services - Requests page
+    """
 
+    REQUEST_FINISHED_STATES = {'Migrated', 'Finished'}
 
-# TODO Refactor these module methods and their callers for a proper request class
-class Request(Navigatable):
-    def __init__(self, appliance=None):
+    def __init__(self, description=None, cells=None, partial_check=False, appliance=None):
+        """
+        Args:
+            description: by default we'll be checking Description column to find required row
+            cells: cells used to find required row in table
+            partial_check: greedy search or not?
+        """
         Navigatable.__init__(self, appliance=appliance)
+        self.description = description
+        self.partial_check = partial_check
+        self.cells = {'Description': self.description} if cells is None else cells
+        self.rest = self._get_request_from_rest(self.partial_check)
+        self.row = None
+
+    @variable(alias='rest')
+    def wait_for_request(self):
+        def _finished():
+            self.rest.reload()
+            if self.rest.request_state.title() not in self.REQUEST_FINISHED_STATES:
+                return False
+            return True
+
+        wait_for(_finished, num_sec=800, delay=20, message="Request finished")
+
+    @wait_for_request.variant('ui')
+    def wait_for_request_ui(self):
+        def _finished():
+            self.update(method='ui')
+            if self.row.request_state.text not in self.REQUEST_FINISHED_STATES:
+                return False
+            return True
+
+        wait_for(_finished, num_sec=800, delay=20, message="Request finished")
+
+    def _get_request_from_rest(self, partial_check):
+        self.appliance.rest_api.collections.requests.reload()
+        requests = self.appliance.rest_api.collections.requests.all
+        matching_requests = []
+        for request in requests:
+            if partial_check:
+                if self.cells['Description'] in request.description:
+                    matching_requests.append(request)
+            else:
+                if self.cells['Description'] == str(request.description):
+                    matching_requests.append(request)
```

Comment:
You can (should) rewrite all the code above that's involved in getting ``matching_requests`` with ``self.appliance.rest_api.collections.requests.find_by(description=self.cells['Description'])`` (or ``description='{}%'.format(self.cells['Description'])`` for ``partial_check``). This will find all matching entities and you also don't need to do the ``reload()``.